metals/src/main/scala/scala/meta/internal/metals/codeactions/ExtractClass.scala

```diff
@@ -0,0 +1,250 @@
+      implicit ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = {
+    val uri = params.getTextDocument.getUri
+    val path = uri.toAbsolutePath
+    val range = params.getRange
+
+    Future.successful {
+      trees.get(path) match {
+        case Some(tree) =>
+          // Get the file name without extension
+          val fileName = uri.toAbsolutePath.filename.replaceAll("\\.scala$", "")
+
+          val classDefinitions = getClassDefinitions(tree)
+
+          val fileAndClassHaveDifferentName = classDefinitions.length == 1 &&
+            classDefinitions.head.name.value != fileName &&
+            classDefinitions.head.name.pos.toLSP.overlapsWith(range) &&
+            !getExtendsNames(classDefinitions.head).contains(fileName)
+
+          // This check is here only to avoid doing more computations in case of a single class in the file
+          if (classDefinitions.length > 1) {
+            getNewFileTreeStructure(tree, range, fileName) match {
+              case Some((newTree, classDefn)) =>
+                extractClassAction(uri, newTree, classDefn) :: Nil
+              case _ => Nil
+            }
+          } else if (fileAndClassHaveDifferentName) {
+            List(renameFileAsClassAction(uri, classDefinitions.head))
+          } else Nil
+
+        case _ => Nil
+      }
+
+    }
+  }
+
+  private def getClassDefinitions(tree: Tree): List[Defn.Class] = {
+    val nodes: ListBuffer[Defn.Class] = ListBuffer()
+
+    val traverser = new SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg =>
+          super.apply(p)
+        case classNode: Defn.Class =>
+          nodes += classNode
+        case s: Source =>
+          super.apply(s)
+        case _ =>
+      }
+    }
+    traverser(tree)
+
+    nodes.toList
+  }
+
+  private def getNewFileTreeStructure(
+      tree: Tree,
+      range: l.Range,
+      currFileName: String
+  ): Option[(Tree, Defn.Class)] = {
+
+    val sealedDefn = tree.collect {
+      case node: Defn.Trait if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+        node.name.value
+      case node: Defn.Class if node.mods.exists(_.isInstanceOf[Mod.Sealed]) =>
+        node.name.value
+    }
+
+    def classExtendsSealed(classDefn: Defn.Class): Boolean =
+      getExtendsNames(classDefn).exists(sealedDefn.contains(_))
+
+    // List of sequential packages or imports before the class definition
+    val structure: ListBuffer[Stat] = ListBuffer()
+    var classFound: Option[Defn.Class] = None
+
+    // Using a custom traverser to avoid hitting inner classes by stopping the recursion on Defn.Class
+    object traverser extends SimpleTraverser {
+      override def apply(tree: Tree): Unit = tree match {
+        case p: Pkg if p.pos.toLSP.overlapsWith(range) =>
+          structure += Pkg(ref = p.ref, stats = Nil)
+          super.apply(p)
+        case i: Import =>
+          structure += i
+        case classNode: Defn.Class
```

Comment:
If we get class calculated previously we can just get use it and only add imports